### PR TITLE
Add Inline JDA Builder

### LINKED ISSUE #8: Add Inline JDA Builder
Add inline JDA builder, so creating the JDA instance can be more idiomatic.

Example usage:
```kotlin
val jda = DefaultJDA(botConfig.token) {
    memberCachePolicy = MemberCachePolicy.ONLINE or MemberCachePolicy.VOICE or MemberCachePolicy.OWNER
    chunkingFilter = ChunkingFilter.ALL
    compression = Compression.ZLIB
    largeThreshold = 250
    
    rawEventsEnabled = false
    enableShutdownHook = true
    bulkDeleteSplittingEnabled = false
    
    eventListeners += LoggingListener(this@PolyBot)
    eventListeners += MessageCacheListener(this@PolyBot)
    
    builder.injectKTX()
}
```

I have the code for this, will pr in a few.

### PR DESCRIPTION
**Pull request by solonovamax**
Closes #8 

Adds an idiomatic way of creating the JDA instance in Kotlin.

Simple to use, looks clean.

### CODE REVIEW

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:165]: Why add deprecated things?

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:287]: The thing is when this event manager is changed and injectKtx is true that will cause the manager to be ignored, maybe there is a better way to handle that?

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:302]: Missing field assignment

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:167]: These inlines don't do anything

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:142]: Shouldn't these defaults properly reflect the state of the builder?

File: src/main/kotlin/dev/minn/jda/ktx/utils.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/utils.kt:16]: Docs?

File: src/main/kotlin/dev/minn/jda/ktx/utils.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/utils.kt:22]: Docs

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:381]: This should not be a nested class

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:383]: This default makes no sense

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:326]: Why not just `sharding(shardId: Int = 0, shardTotal: Int = 1)`

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:1]: Copyright header

File: src/main/kotlin/dev/minn/jda/ktx/utils.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/utils.kt:1]: Copyright header

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:391]: This should not be here

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:389]: Why can you set them here? The best way to configure intents and cache should be through `enableIntents` and `enableCache` rather than setters.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:387]: The thing is, when you build multiple times this will screw you over.

**Review by MinnDevelopment — CHANGES_REQUESTED** [review]
I'm unsure whether this is a useful feature to add

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:287]: Hmmm, fair point.

Here's perhaps another idea:
rewrite the injected manager slightly so that it takes an event manager in the constructor. Then, it will delegate the event registration to the new even manager and if it fails, it sends registers it on its own.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:142]: Perhaps they should, but they can also change at any time. But when they're set to null, it preserves the default value.

File: src/main/kotlin/dev/minn/jda/ktx/utils.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/utils.kt:16]: Just wanted to explain why this was added.

Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/utils.kt:16]: wait do you _want_ docs or not want them?

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:381]: I was following [`builder.kt`](https://github.com/MinnDevelopment/jda-ktx/blob/master/src/main/kotlin/dev/minn/jda/ktx/builder.kt#L274), as that uses nested classes.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:383]: By default, JDA only uses 1 shard with an id of 0 (Actually, should that `shardTotal` be `1`? I might have overlooked that)
but if anyone _is_ using that, then they basically _have_ to assign both of those values. The default is just to get around the compiler warnings. Though, `lateinit` could be used, but I'm not sure how you feel about that.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:1]: oops, forgot that.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:391]: Where should it be moved to then? I added it because it's build logic for the wrapped builder, so it'd intuitively go in the build method.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:389]: I chose to add both `setEnabledIntents` _and_ `enableIntents`. it's recommended to use the `enableIntents` methods that are provided, but if you want to use `setEnabledIntents`, then the variable `gatewayIntents` is provided for that. By default it's initialized to an empty set, so it's ignored.

I would have *much* preferred to just check which intents are enabled from the builder, and then use that as a backing field, but I didn't want to put in any reflection hacks.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:387]: Ah, that is a _very_ fair point.

I'm gonna try and redo this a bit for sharding.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:165]: Originally I was very unsure of how to enable access to the cache flags. I wanted to provide a field for it because that just seemed cleaner to me, but in the end I gave up with that and decided to just have the methods.
I *could* make a custom class that overloads the `plusAssign` and `minusAssign` operators and uses those to add them, though.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt]: I wasn't sure, since I don't know exactly how kotlin `inline` works lol

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:401]: Looking at some more stuff, it might be better to just provide a builder for `DefaultShardManager`.
Because if you're using a single shard, it functions the same way as a normal JDA instance.
Although, there might be some reasons that one might want to construct a `JDA` instance instead of a `DefaultShardManager`.

I'm open to suggestions.

File: src/main/kotlin/dev/minn/jda/ktx/utils.kt
Comment (MinnDevelopment) [reply src/main/kotlin/dev/minn/jda/ktx/utils.kt:33]: I want docs but not just `@param name` which is rather useless

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:401]: I think it would be better to just use a method with default values like i suggested in the other comment. The pattern here doesn't really make sense here since you just set an id essentially.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:409]: I mean that using these setters should be avoided in general and instead the `enableCache` and `enableIntents` methods should be used instead. The setters will replace the defaults which is a footgun.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:401]: I saw that.

Actually, thinking about it, it might be better to just have a single method, without any default parameters, that takes in both of the values for it.

Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:401]: Actually actually, it might even be a bit nicer if we instead use a variable of type `Pair<Int, Int>` and it'd be used like this:

```kotlin
val builder = DefaultJDABuilder("token") {
    // stuff
}

for (id in 0..10) {
    builder.sharding = id to 10
    builder.build()
}
```

File: src/main/kotlin/dev/minn/jda/ktx/utils.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/utils.kt:33]: aight, will do.

I just left that there because intellij auto generate stuff.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:409]: Yeah, I know.
They should probably be both annotated with `@Deprecated`, and have a warning in the javadoc saying not to use the `gatewayIntents` or `cacheFlags` variables. But I still added that in, because it's functionality that the builder has, and I don't want to restrict anything.

**Comment by d1snin** [comment]
we write functions in camelCase, they should not start with a capital letter. Or is it a new style that I don't know about? I haven't seen this in dsl yet

**Comment by solonovamax** [comment]
@d1snin See: [`fun EmbedBuilder`](https://github.com/MinnDevelopment/jda-ktx/blob/master/src/main/kotlin/dev/minn/jda/ktx/builder.kt#L106), [`fun MessageBuilder`](https://github.com/MinnDevelopment/jda-ktx/blob/master/src/main/kotlin/dev/minn/jda/ktx/builder.kt#L84), [`fun Embed`](https://github.com/MinnDevelopment/jda-ktx/blob/master/src/main/kotlin/dev/minn/jda/ktx/builder.kt#L53), and [`fun Message`](https://github.com/MinnDevelopment/jda-ktx/blob/master/src/main/kotlin/dev/minn/jda/ktx/builder.kt#L30)

**Comment by solonovamax** [comment]
It's used because these are kind of meant to just look like constructors.
I've seen some people who use it and others who don't. I'm not sure where I stand on it just yet, so I went with what had already beed established by Minn as a "standard" for this repo.

**Comment by d1snin** [comment]
this is of course your choice, but as a dsl fan, it would be nice for me to use these functions when they were written in camelCase, otherwise I would really perceive them as constructors 😄

**Comment by solonovamax** [comment]
I'll leave it up to Minn.
I'm impartial to it either way. (Though, it does have a certain style to it in CamelCase as opposed to camelCase.)

**Comment by d1snin** [comment]
okay, no problem, but this is the first time I've seen this tbh

**Comment by MinnDevelopment** [comment]
The naming pattern is supposed to represent them as initializer functions similar to [`List(size) { index -> element }`](https://kotlinlang.org/docs/constructing-collections.html#initializer-functions-for-lists).

**Comment by solonovamax** [comment]
Resolved all the changes + added builder for DefaultShardManager.

File: src/main/kotlin/dev/minn/jda/ktx/DelegatingCollection.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/DelegatingCollection.kt:26]:
```suggestion
) : MutableCollection<T> by delegate {
```

File: src/main/kotlin/dev/minn/jda/ktx/CoroutineEventManager.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/CoroutineEventManager.kt:41]:
```suggestion
) : IEventManager {
```

File: src/main/kotlin/dev/minn/jda/ktx/CoroutineEventManager.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/CoroutineEventManager.kt:63]:
```suggestion

```

File: src/main/kotlin/dev/minn/jda/ktx/CoroutineEventManager.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/CoroutineEventManager.kt:61]:
```suggestion

```

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:58]:
```suggestion
) = InlineShardManagerBuilder(DefaultShardManagerBuilder.createDefault(token))
```

Comment (MinnDevelopment) [reply src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:58]: And all the other instances of this. Please adhere to my codestyle.

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:80]: This is unreadable, please change it.

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:244]: I still think these should represent the correct defaults that are applied to the builders

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:317]: Remove things that are deprecated

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:365]: I think it would be better to have a pattern like this:

```kotlin
status(ONLINE)

status { 
  if (it == 1) ONLINE
  else OFFLINE
}
```

instead of the current property syntax which assigns functions you have here.

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:397]: Better:

```kotlin
rateLimitPool(false, scheduler)

rateLimitPool(false) {
  newPool()
}
```

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:364]:
```suggestion
    )
```

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:378]: I think you should remove the property `sharding = 1 to 2` doesn't look very nice regardless. You aren't updating the sharding field here either, better just remove it entirely.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:398]: I think this is unused?

File: src/main/kotlin/dev/minn/jda/ktx/inject.kt
Comment (MinnDevelopment) [inline src/main/kotlin/dev/minn/jda/ktx/inject.kt:29]: Why not just set this to the actual default instead of null? This makes it harder to understand for the library user

Comment (MinnDevelopment) [reply src/main/kotlin/dev/minn/jda/ktx/inject.kt:29]: There is no explanation what "delegate" is here. Please add some documentation since this is non-obvious.

**Review by MinnDevelopment — CHANGES_REQUESTED** [review]

File: src/main/kotlin/dev/minn/jda/ktx/inject.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/inject.kt:29]: will do.

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:58]: There was no included `.editorconfig`, so I use assumed my own code style as it wasn't /too/ different from yours. But, will do.

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:244]: The only problem is, that the defaults within the builder are subject to change. There is no reason that they will remain the same across versions. And since the defaults are not accessible except via reflection, there is no nice way to guarantee parity with the JDA builder defaults.

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:317]: It is deprecated, but still supported as a feature. Therefore, I left it in.

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:365]: Hmmm. The second one looks nice.
What about the following?
```kotlin
status = ONLINE

status {
    if (it == 1)
        ONLINE
    else
        OFFLINE
}
```

File: src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/InlineShardManagerBuilder.kt:397]: While making this I forgot about all about kotlin's fancy lambda functions. Will add those changes to everything.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:378]: I agree.

File: src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt
Comment (solonovamax) [reply src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt:392]: oops lol

**Comment by solonovamax** [comment]
@MinnDevelopment Regarding the default values:

For the default values, I left them as `null` because the JDABuilder defaults are subject to change without notice. Therefore, in order to make sure that it always uses the default from the `JDABuilder` class, they must be null, as that then uses the default value.
The only way to use the JDA defaults reliably would be via reflection, which I think should be avoided.

**Comment by solonovamax** [comment]
ay, I don't wanna be annoying or anything, but do you have like a discord we could talk about this in or smth? It'd be easier than through gh comments

**Comment by solonovamax** [comment]
@MinnDevelopment don't want to be pushy, but any updates?

**Comment by solonovamax** [comment]
Any updates? I'm waiting for this to be merged so I can use it in my bot lol

**Comment by RyanMalaspina** [comment]
> Any updates? I'm waiting for this to be merged so I can use it in my bot lol

[Minn announced he's stepping away from JDA](https://discord.com/channels/125227483518861312/169489002708074496/870395911409397771) so I assume that means this as well

**Comment by solonovamax** [comment]
I'm going to be closing this and maintaining my own fork of jda-ktx for the time being.
If you'd like to merge, I'm more than happy to do that.

**Pull request closed without merging**
